### PR TITLE
NEW: more tests for editor

### PR DESCRIPTION
**Pull request by deepansh96**
Fixes #{issue id}

## Test Plan

- [ ] Test Responsiveness
   - [ ] Laptop (1200px)
   - [ ] Tablet (760px)
   - [ ] Phone (320px)
- [ ] Cross-Browser Testing
   - [ ] Chrome
   - [ ] Firefox
   - [ ] Safari
- [ ] Local Language Support
- [ ] Wrote tests
- [ ] Tested locally
- [ ] Tested on staging
- [ ] Tested on an actual physical phone
- [ ] Tested on production
- [ ] Check for bundle size [here](https://bundlephobia.com/) if adding a package
- [ ] Added relevant details like Labels/Projects/Milestones etc.

### CODE REVIEW

File: src/components/UI/Alert/DialogBox.vue [hunk 1]
```diff
@@ -117,6 +117,8 @@ export default {
       // is either of the confirm or cancel buttons visible
       // the only way to hide each of those two buttons is by
       // adding the `hidden` class to each of those button classes
+      if (this.confirmButtonClass == undefined || this.cancelButtonClass == undefined)
```
Comment (dalmia) [inline src/components/UI/Alert/DialogBox.vue:120]: why was this added?

File: src/pages/Editor.vue [hunk 2]
```diff
@@ -553,16 +572,11 @@ export default {
         "hover:bg-primary hover:text-white hover:border-primary": !this.addItemDisabled,
       };
     },
-    currentItemType() {
```
Comment (dalmia) [inline src/pages/Editor.vue:556]: was this not being used?

File: src/pages/Editor.vue [hunk 3]
```diff
@@ -638,10 +652,6 @@ export default {
         invalidMessage: this.$t("editor.video_input.validation.invalid"),
       };
     },
-    urlStyleClass() {
```
Comment (dalmia) [inline src/pages/Editor.vue:641]: was this not being used?

File: src/pages/Editor.vue [hunk 4]
```diff
@@ -695,14 +705,6 @@ export default {
...
       };
     },
-    backButtonIconConfig() {
-      // config for icon of back button
-      return {
-        enabled: true,
-        iconName: "chevron-left-solid",
-        iconClass: "w-4 h-4 ml-2 text-primary",
-      };
-    },
```
Comment (dalmia) [inline src/pages/Editor.vue:705]: was this not being used?

File: src/pages/Editor.vue [hunk 5]
```diff
@@ -811,17 +809,6 @@ export default {
...
-      // styling class for add item button
-      // disabled the button if plio is published
-      var classObject = [
-        { "cursor-not-allowed": this.addItemDisabled },
-
-        `rounded-md font-bold p-5 h-12 w-full bg-primary-button ring-primary
-        hover:bg-primary-button-hover disabled:opacity-50 shadow-lg`,
-      ];
-      return classObject;
-    },
```
Comment (dalmia) [inline src/pages/Editor.vue:824]: was this not being used?

File: src/pages/Editor.vue [hunk 6]
```diff
@@ -759,10 +761,6 @@ export default {
       // whether the plio has been pubished
       return this.status == "published";
     },
-    isDraftCreated() {
-      // whether the draft has been created
-      return this.plioId != "";
-    },
```
Comment (dalmia) [inline src/pages/Editor.vue:765]: was this not being used?

File: src/pages/Editor.vue [hunk 7]
```diff
@@ -1278,7 +1261,6 @@ export default {
     addNewItem(questionType) {
       this.player.pause();
       this.startLoading();
-      setTimeout(() => {}, 5000);
```
Comment (dalmia) [inline src/pages/Editor.vue:1281]: why was this removed?

File: src/services/Testing/DummyData.js [hunk 8]
```diff
@@ -285,6 +285,25 @@ export const dummyItems = {
   ],
 };
 
+export const imageData = {
+  dataUrl:
```
Comment (dalmia) [inline src/services/Testing/DummyData.js:289]: did you compute this value for an image and then place it here?

File: src/services/Testing/DummyData.js [hunk 9]
```diff
@@ -285,6 +285,25 @@ export const dummyItems = {
...
+  info: {
+    name: "Screenshot 2021-07-08 at 11.26.42 PM.png",
```
Comment (dalmia) [inline src/services/Testing/DummyData.js:292]: can we make the name as "dummy"?

File: tests/unit/pages/Editor.spec.js [hunk 10]
```diff
@@ -2,7 +2,23 @@ import { mount, flushPromises } from "@vue/test-utils";
...
+  imageData,
+} from "@/services/Testing/DummyData.js";
+import store from "@/store";
+
+var cloneDeep = require("lodash.clonedeep");
+let dummyItemsCopy;
+
+beforeEach(() => {
+  jest.useFakeTimers();
+  dummyItemsCopy = cloneDeep(dummyItems);
```
Comment (dalmia) [inline tests/unit/pages/Editor.spec.js:20]: why create this copy?

File: tests/unit/pages/Editor.spec.js [hunk 11]
```diff
@@ -103,4 +121,867 @@ describe("Editor.vue", () => {
...
+    // setInterval would've been called again after 5 seconds
+    // but as `changeInProgress` is false, `savePlio` will not be called
+    jest.advanceTimersByTime(timeInterval);
+    expect(savePlio).not.toHaveBeenCalled();
+
+    // change `changeInProgress` to true,
+    // and check before & after 5 seconds
+    await wrapper.setData({ changeInProgress: true });
+    expect(savePlio).not.toHaveBeenCalled();
+    jest.advanceTimersByTime(timeInterval);
```
Comment (dalmia) [inline tests/unit/pages/Editor.spec.js:146]: nice :P

File: tests/unit/pages/Editor.spec.js [hunk 12]
```diff
@@ -103,4 +121,867 @@ describe("Editor.vue", () => {
...
+    expect(checkAndSavePlio).not.toHaveBeenCalled();
+
+    // add items to the component, the method should've been called
+    await wrapper.setData({ items: dummyItemsCopy.data });
+    expect(checkAndSavePlio).toHaveBeenCalled();
+
+    // update the items, method should've been called
+    let updatedDummyItems = dummyItemsCopy.data;
+    updatedDummyItems[0].time = 20;
+    await wrapper.setData({ items: updatedDummyItems });
```
Comment (dalmia) [inline tests/unit/pages/Editor.spec.js:164]: this would work the same way right?
```suggestion
    dummyItemsCopy.data[0].time = 20;
    await wrapper.setData({ items: dummyItemsCopy.data });
```

File: tests/unit/pages/Editor.spec.js [hunk 13]
```diff
@@ -103,4 +121,867 @@ describe("Editor.vue", () => {
...
+  });
+
+  it("handles title updation correctly", async () => {
+    const checkAndSavePlio = jest.spyOn(Editor.methods, "checkAndSavePlio");
+    const wrapper = mount(Editor);
+
+    await wrapper.setData({ plioTitle: "title for plio" });
+    expect(wrapper.vm.loadedPlioDetails.plioTitle).not.toBe(
+      wrapper.vm.plioTitle
+    );
```
Comment (dalmia) [inline tests/unit/pages/Editor.spec.js:346]: shouldn't we also check equality with the newly set title value?

File: tests/unit/pages/Editor.spec.js [hunk 14]
```diff
@@ -103,4 +121,867 @@ describe("Editor.vue", () => {
...
+
+    const imageURL = "test url";
+    const dummyItemsWithImage = dummyItemsCopy.data;
+    dummyItemsWithImage[0].details.image = {
+      id: 56,
+      url: imageURL,
+      alt_text: "Image",
+      created_at: "2021-07-02T12:58:41.683683Z",
+      updated_at: "2021-07-02T12:58:41.684174Z",
+    };
```
Comment (dalmia) [inline tests/unit/pages/Editor.spec.js:368]: shouldn't this use the dummyImage object that you imported at the start?

File: tests/unit/pages/Editor.spec.js [hunk 15]
```diff
@@ -103,4 +121,867 @@ describe("Editor.vue", () => {
...
+    await wrapper.setData({
+      status: "published",
+      hasUnpublishedChanges: true,
+    });
+    expect(wrapper.vm.publishButtonTooltip).toBe(
+      "Click to publish your changes"
+    );
+    expect(
+      wrapper.find('[data-test="publishButton"]').element.$_ptooltipValue
+    ).toBe("Click to publish your changes");
```
Comment (dalmia) [inline tests/unit/pages/Editor.spec.js:412]: can we load this from the language file and read it from there?

File: tests/unit/pages/Editor.spec.js [hunk 16]
```diff
@@ -103,4 +121,867 @@ describe("Editor.vue", () => {
...
+      mockAxios.lastReqGet().data
+    );
+
+    mockAxios.mockResponse(
+      {
+        data: "mock response",
+      },
+      mockAxios.lastReqGet()
+    );
+
```
Comment (dalmia) [inline tests/unit/pages/Editor.spec.js:670]: shouldn't we also check if `/items` is being updated with the image?

File: tests/unit/pages/Editor.spec.js [hunk 17]
```diff
@@ -103,4 +121,867 @@ describe("Editor.vue", () => {
...
+      videoId: "jdYJf_ybyVo",
+      currentTimestamp: 15.6,
+    });
+    await store.dispatch("sync/stopLoading");
+
+    await wrapper.find('[data-test="addMCQItem"]').trigger("click");
+    expect(addNewItem).toHaveBeenCalled();
+    expect(showCannotAddItemDialog).toHaveBeenCalled();
+    expect(wrapper.vm.pending).toBeFalsy();
+
```
Comment (dalmia) [inline tests/unit/pages/Editor.spec.js:839]: one comment as to why someone should not be able to add item

File: tests/unit/pages/Editor.spec.js [hunk 18]
```diff
@@ -103,4 +121,867 @@ describe("Editor.vue", () => {
...
+    await wrapper.setData({
+      items: dummyItems.data,
+      currentItemIndex: null,
+      videoId: "jdYJf_ybyVo",
+      currentTimestamp: 12,
+      plioDBId: 13,
+    });
+    await store.dispatch("sync/stopLoading");
```
Comment (dalmia) [inline tests/unit/pages/Editor.spec.js:847]: one comment as to why someone would be able to add an item now.

**Comment by dalmia** [comment]
4 test cases are failing - please fix them and update.

File: src/services/Testing/DummyData.js (hunk 8 above)
Comment (deepansh96) [reply src/services/Testing/DummyData.js:289]: Yep

File: src/pages/Editor.vue (hunk 7 above)
Comment (deepansh96) [reply src/pages/Editor.vue:1280]: Kuch use hi nhi dikha iska. It was just waiting 5 seconds

File: src/pages/Editor.vue (hunk 5 above)
Comment (deepansh96) [reply src/pages/Editor.vue:824]: Nope

File: src/pages/Editor.vue (hunk 6 above)
Comment (deepansh96) [reply src/pages/Editor.vue:765]: Nope

File: src/pages/Editor.vue (hunk 4 above)
Comment (deepansh96) [reply src/pages/Editor.vue:705]: Nope

File: src/pages/Editor.vue (hunk 2 above)
Comment (deepansh96) [reply src/pages/Editor.vue:556]: At one place but there was a duplicate computed property called `itemType` so used that instead

File: src/components/UI/Alert/DialogBox.vue (hunk 1 above)
Comment (deepansh96) [reply src/components/UI/Alert/DialogBox.vue:120]: If for some reason `confirmButtonClass` or `cancelButtonClass` was undefined, then the code below, `this.confirmButtonClass.indexOf("hidden")` gave an error like `undefined has no method indexOf`

File: tests/unit/pages/Editor.spec.js (hunk 13 above)
Comment (deepansh96) [reply tests/unit/pages/Editor.spec.js:372]: I mean we're passing a title as setData to `plioTitle` only, so wouldn't make much sense

File: tests/unit/pages/Editor.spec.js (hunk 15 above)
Comment (deepansh96) [reply tests/unit/pages/Editor.spec.js:432]: Not sure
In all of our testing code we have done this way only

File: tests/unit/pages/Editor.spec.js (hunk 16 above)
Comment (deepansh96) [reply tests/unit/pages/Editor.spec.js:655]: That is being checked
